### PR TITLE
add DPM scheduler with EDM formulation

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 1]
```diff
@@ -0,0 +1,676 @@
...
+        euler_at_final (`bool`, defaults to `False`):
+            Whether to use Euler's method in the final step. It is a trade-off between numerical stability and detail
+            richness. This can stabilize the sampling of the SDE variant of DPMSolver for small number of inference
+            steps, but sometimes may result in blurring.
+        final_sigmas_type (`str`, defaults to `"zero"`):
+            The final `sigma` value for the noise schedule during the sampling process. If `"sigma_min"`, the final sigma
+            is the same as the last sigma in the training schedule. If `zero`, the final sigma is set to 0.
+    """
+
+    _compatibles = [e.name for e in KarrasDiffusionSchedulers]
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:84]: should this be empty for now?

**Review by sayakpaul — COMMENTED** [review]
Generally looks good. Let's make sure to add "Copied from ..." statements wherever applicable.

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 2]
```diff
@@ -0,0 +1,676 @@
...
+        lower_order_final: bool = True,
+        euler_at_final: bool = False,
+        final_sigmas_type: Optional[str] = "zero",  # "zero", "sigma_min"
+    ):
+        # settings for DPM-Solver
+        if algorithm_type not in ["dpmsolver++", "sde-dpmsolver++"]:
+            if algorithm_type == "deis":
+                self.register_to_config(algorithm_type="dpmsolver++")
+            else:
+                raise NotImplementedError(f"{algorithm_type} does is not implemented for {self.__class__}")
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:111]:
```suggestion
                raise NotImplementedError(f"{algorithm_type} is not implemented for {self.__class__}")
```

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+
+        if solver_type not in ["midpoint", "heun"]:
+            if solver_type in ["logrho", "bh1", "bh2"]:
+                self.register_to_config(solver_type="midpoint")
+            else:
+                raise NotImplementedError(f"{solver_type} does is not implemented for {self.__class__}")
+
+        if algorithm_type not in ["dpmsolver++", "sde-dpmsolver++"] and final_sigmas_type == "zero":
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:119]: should `deis` be accepted here as well?

**Review by pcuenca — APPROVED** [review]
Looks good in principle!

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py (hunk 3 above)
Comment (patil-suraj) [reply src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:119]: Not sure, for now I followed the logic in existing dpm scheduler.

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:119]: I think we just map it to `dpmsolver++` here so it should be fine; although I absolutely have no clue why they are there (I looked back at the original PR that added it too)

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 4]
```diff
@@ -0,0 +1,683 @@
...
+        w = (low - log_sigma) / (low - high)
+        w = np.clip(w, 0, 1)
+
+        # transform interpolation to time range
+        t = (1 - w) * low_idx + w * high_idx
+        t = t.reshape(sigma.shape)
+        return t
+
+    def _sigma_to_alpha_sigma_t(self, sigma):
+        alpha_t = torch.tensor(1)  # Inputs are pre-scaled before going into unet, so alpha_t = 1
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:338]: ohh nice! that's why this formula still works here 
```
 x_t = (sigma_t / sigma_s) * sample - (alpha_t * (torch.exp(-h) - 1.0)) * model_output
```
or maybe we do not need this function at all , just change the math directly inside the steps as

```
 x_t = (sigma_t / sigma_s) * sample -  (torch.exp(-h) - 1.0) * model_output
```

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 5]
```diff
@@ -0,0 +1,683 @@
...
+            `torch.FloatTensor`:
+                The sample tensor at the previous timestep.
+        """
+        sigma_t, sigma_s = self.sigmas[self.step_index + 1], self.sigmas[self.step_index]
+        alpha_t, sigma_t = self._sigma_to_alpha_sigma_t(sigma_t)
+        alpha_s, sigma_s = self._sigma_to_alpha_sigma_t(sigma_s)
+        lambda_t = torch.log(alpha_t) - torch.log(sigma_t)
+        lambda_s = torch.log(alpha_s) - torch.log(sigma_s)
+
+        h = lambda_t - lambda_s
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:403]:
```suggestion
        h = torch.log(sigma_s) - torch.log(sigma_t)
```

I'm not 100% sure the math is correct here but the idea is we can calculate `h` from sigma directly

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 6]
```diff
@@ -0,0 +1,683 @@
...
+        if self.config.algorithm_type == "dpmsolver++":
+            x_t = (sigma_t / sigma_s) * sample - (alpha_t * (torch.exp(-h) - 1.0)) * model_output
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:405]:
```suggestion
            x_t = (sigma_t / sigma_s) * sample - (torch.exp(-h) - 1.0) * model_output
```

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 7]
```diff
@@ -0,0 +1,683 @@
...
+        elif self.config.algorithm_type == "sde-dpmsolver++":
+            assert noise is not None
+            x_t = (
+                (sigma_t / sigma_s * torch.exp(-h)) * sample
+                + (alpha_t * (1 - torch.exp(-2.0 * h))) * model_output
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:410]:
```suggestion
                + (1 - torch.exp(-2.0 * h)) * model_output
```

**Review by yiyixuxu — COMMENTED** [review]
look good!
I left a comment about the `_sigma_to_alpha_sigma_t` function - I think maybe we don't need this here and can simplify the math a little bit

File: src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py [hunk 8]
```diff
@@ -0,0 +1,683 @@
...
+            `torch.FloatTensor`:
+                The sample tensor at the previous timestep.
+        """
+        sigma_t, sigma_s0, sigma_s1 = (
+            self.sigmas[self.step_index + 1],
+            self.sigmas[self.step_index],
+            self.sigmas[self.step_index - 1],
+        )
+
+        alpha_t, sigma_t = self._sigma_to_alpha_sigma_t(sigma_t)
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_edm_dpmsolver_multistep.py:441]: same comments as https://github.com/huggingface/diffusers/pull/7120#discussion_r1504724425
